File: src/com/codingspezis/android/metalonly/player/favorites/SongSaver.java
package com.codingspezis.android.metalonly.player.favorites;

import java.util.*;

import android.content.*;

import com.codingspezis.android.metalonly.player.*;

/**
 * 
 * this saves a list of songs to shared preferences
 * 
 * TODO: get rid of the quick and dirty solution of saving songs via shared
 * preferences -> do this with XML files
 * 
 */
public class SongSaver {

	// part two of shared preferences key
	public static final String KEY_INTERPRET = "_INTERPRET_";
	public static final String KEY_TITLE = "_TITLE_";
	public static final String KEY_THUMB = "_THUMB_";
	public static final String KEY_DATE = "_DATE_";
	public static final String KEY_COUNT = "_COUNT";

	// given by constructor
	private final String sharedPreferencesPrefix;
	private int limit;

	// holding
	private LinkedList<Song> songList;
	private boolean changes;
	private final SharedPreferences sharedPreferences;

	/**
	 * constructor
	 * 
	 * @param context
	 *            context
	 * @param sharedPreferencesPrefix
	 *            prefix to use for saving songs in shared preferences
	 */
	public SongSaver(Context context, String sharedPreferencesPrefix, int limit) {
		this.sharedPreferencesPrefix = sharedPreferencesPrefix;
		if (limit < 0) {
			this.limit = Integer.MAX_VALUE;
		} else {
			this.limit = limit;
		}
		sharedPreferences = context.getSharedPreferences(
				context.getString(R.string.app_name),
				Context.MODE_MULTI_PROCESS);
		readSongsFromStorage();
		changes = false;
	}

	/**
	 * reads songs from storage
	 * 
	 * @return list of songs from storage
	 */
	private void readSongsFromStorage() {
		int numberOfSongs = sharedPreferences.getInt(sharedPreferencesPrefix
				+ KEY_COUNT, 0);
		songList = new LinkedList<Song>();
		for (int i = 0; i < numberOfSongs; i++) {

			String interpret = sharedPreferences.getString(
					sharedPreferencesPrefix + KEY_INTERPRET + i, "");
			String title = sharedPreferences.getString(sharedPreferencesPrefix
					+ KEY_TITLE + i, "");
			String thumb = sharedPreferences.getString(sharedPreferencesPrefix
					+ KEY_THUMB + i, "");
			long date = sharedPreferences.getLong(sharedPreferencesPrefix
					+ KEY_DATE + i, 0);

			Song song = new Song(interpret, title, thumb, date);

			if (!addSong(song)) {
				break;
			}
		}
	}

	/**
	 * saves songs to storage
	 */
	public void saveSongsToStorage() {
		if (somethingChanged()) {
			SharedPreferences.Editor editor = sharedPreferences.edit();
			for (int i = 0; i < songList.size(); i++) {
				final Song song = songList.get(i);
				editor.putString(sharedPreferencesPrefix + KEY_INTERPRET + i,
						song.interpret);
				editor.putString(sharedPreferencesPrefix + KEY_TITLE + i,
						song.title);
				editor.putString(sharedPreferencesPrefix + KEY_THUMB + i,
						song.getThumb());
				editor.putLong(sharedPreferencesPrefix + KEY_DATE + i,
						song.date);
			}
			editor.putInt(sharedPreferencesPrefix + KEY_COUNT, songList.size());
			editor.commit();
		}
	}

	/**
	 * adds song to list of songs
	 * 
	 * @param song
	 *            song to add
	 * @return true if adding was successful - false otherwise
	 */
	public boolean addSong(Song song) {
		if (song.isValid() && isAlreadyIn(song) == -1
				&& limit > songList.size()) {
			songList.add(song);
			changes = true;
			return true;
		}
		return false;
	}

	/**
	 * adds song to song list - if limit is reached first in will be deleted
	 * 
	 * @param song
	 *            song to add
	 */
	public boolean queeIn(Song song) {
		while (limit <= songList.size()) {
			songList.remove();
		}
		return addSong(song);
	}

	/**
	 * checks if song is already in list of songs
	 * 
	 * @param song
	 *            song to check
	 * @return index of last entry of song if it is in the list - -1 otherwise
	 */
	public int isAlreadyIn(Song song) {
		for (int i = songList.size() - 1; i >= 0; i--) {
			if (songList.get(i).interpret.equals(song.interpret)
					&& songList.get(i).title.equals(song.title)) {
				return i;
			}
		}
		return -1;
	}

	/**
	 * removes song at position i
	 * 
	 * @param i
	 *            position of song to delete
	 */
	public void removeAt(int i) {
		changes = true;
		songList.remove(i);
	}

	/**
	 * removes every song of the list
	 */
	public void clear() {
		while (!songList.isEmpty()) {
			changes = true;
			songList.remove();
		}
	}

	/**
	 * getter for ith song of list of songs
	 * 
	 * @param i
	 *            index of requested song
	 * @return ith song of list of songs
	 */
	public Song get(int i) {
		return songList.get(i);
	}

	/**
	 * @return true if there were changes on list of songs since they have been
	 *         read - false otherwise
	 */
	public boolean somethingChanged() {
		return changes;
	}

	/**
	 * @return returns the size of list of songs
	 */
	public int size() {
		return songList.size();
	}

	/**
	 * reloads song list
	 */
	public void reload() {
		readSongsFromStorage();
	}
}

File: app/src/main/kotlin/com/codingspezis/android/metalonly/player/favorites/SongSaver.kt
package com.codingspezis.android.metalonly.player.favorites

import android.content.Context

import com.codingspezis.android.metalonly.player.core.HistoricTrack

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.Collections
import java.util.LinkedList

/**
 * This class persists songs in the internal storage via JSON files.
 * @param context context of calling object
 * @param fileName name of file where to save songs
 * @param limit maximum of songs (FIFO) - if value <= 0 songs are unlimited

 */
class SongSaver(
    private val context: Context,
    private val fileName: String,
    limit: Int
) {
    private var limit: Int = 0
    private val trackList: LinkedList<HistoricTrack> // main data
    private var changes: Boolean = false // is there share to save?

    init {
        if (limit <= 0)
            this.limit = Integer.MAX_VALUE
        else
            this.limit = limit
        trackList = LinkedList<HistoricTrack>()
        readSongsFromStorage()
        changes = false
    }

    /**
     * TODO refactor, we should use jackson directly (i.e. parse file directly into pojo)
     * reads songs from file
     */
    @Synchronized private fun readSongsFromStorage() {

        val isr: InputStreamReader
        try {
            isr = InputStreamReader(context.openFileInput(fileName))
            var s = ""
            var read: Int
            val BUFF_SIZE = 256
            val buffer = CharArray(BUFF_SIZE)
            do {
                read = isr.read(buffer, 0, BUFF_SIZE)
                if (read > 0) s += String(buffer, 0, read)
            } while (read == BUFF_SIZE)
            isr.close()
            val jObj = JSONObject(s)
            val jSongs = jObj.getJSONArray(JSON_ARRAY_SONGS)
            trackList.clear()
            for (i in 0..jSongs.length() - 1) {
                val jSong = jSongs.getJSONObject(i)
                val interpret = jSong.getString(JSON_STRING_INTERPRET)
                val title = jSong.getString(JSON_STRING_TITLE)
                val thumb = jSong.getString(JSON_STRING_THUMB)
                val date = jSong.getLong(JSON_LONG_DATE)
                val track = HistoricTrack(interpret, title, thumb, date)
                queueIn(track)
            }
        } catch (e: FileNotFoundException) {
            // everything is fine - just nothing saved
        } catch (e: IOException) {
            // TODO: error handling (but this should be dead code)
            e.printStackTrace()
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

    /**
     * saves songs to file
     */
    @Synchronized fun saveSongsToStorage() {
        if (somethingChanged()) {
            try {
                val jSongs = JSONArray()
                for (i in trackList.indices) {
                    val track = trackList[i]
                    val jSong = JSONObject()
                    jSong.put(JSON_STRING_INTERPRET, track.artist)
                    jSong.put(JSON_STRING_TITLE, track.title)
                    jSong.put(JSON_STRING_THUMB, track.moderator)
                    jSong.put(JSON_LONG_DATE, track.playedAtAsLong)
                    jSongs.put(jSong)
                }
                val jObj = JSONObject()
                jObj.put(JSON_ARRAY_SONGS, jSongs)

                context.deleteFile(fileName)
                val osr = OutputStreamWriter(context.openFileOutput(fileName, Context.MODE_PRIVATE))
                osr.write(jObj.toString())
                osr.close()
            } catch (e: JSONException) {
                // TODO: error handling (but this should be dead code)
                e.printStackTrace()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    /**
     * adds track to track list

     * @param track track to add
     * *
     * @return true if adding was successful - false otherwise
     */
    fun addSong(track: HistoricTrack): Boolean {
        if (track.isValid &&
                isAlreadyIn(track) == -1 &&
                limit > trackList.size) {
            trackList.add(track)
            changes = true
            return true
        }
        return false
    }

    /**
     * adds track to track list - if limit is reached first in will be deleted

     * @param track track to add
     */
    fun queueIn(track: HistoricTrack): Boolean {
        while (limit <= trackList.size) {
            trackList.remove()
        }
        return addSong(track)
    }

    /**
     * checks if track is already in list of songs

     * @param track track to check
     * *
     * @return index of last entry of track if it is in the list - -1 otherwise
     * *
     */
    @Deprecated("Move to {@link SongSaver#contains(HistoricTrack)}")
    fun isAlreadyIn(track: HistoricTrack): Int {
        for (i in trackList.indices.reversed()) {
            if (trackList[i].artist == track.artist && trackList[i].title == track.title)
                return i
        }
        return -1
    }

    /**
     * Checks if the given track is already in the list of songs
     * @param track track to check
     * *
     * @return `true`, if the song is known. `false` otherwise
     */
    operator fun contains(track: HistoricTrack): Boolean {
        return isAlreadyIn(track) != -1
    }

    /**
     * removes song at position i

     * @param i position of song to delete
     */
    fun removeAt(i: Int) {
        changes = true
        trackList.removeAt(i)
    }

    /**
     * removes every song of the list
     */
    fun clear() {
        while (!trackList.isEmpty()) {
            changes = true
            trackList.remove()
        }
    }

    /**
     * getter for i-th song of list of songs

     * @param i index of requested song
     * *
     * @return i-th song of list of songs
     */
    operator fun get(i: Int): HistoricTrack {
        return trackList[i]
    }

    /**
     * Returns a copy of the historic song list.
     */
    val all: List<HistoricTrack>
        get() {
            val historicTracks = LinkedList(trackList)
            Collections.sort(historicTracks) { (_, _, _, lhs), (_, _, _, rhs) -> (-(lhs - rhs)).toInt() }
            return historicTracks
        }

    /**
     * @return true if there were changes on list of songs since they have been read - false otherwise
     */
    fun somethingChanged(): Boolean {
        return changes
    }

    /**
     * @return returns the size of list of songs
     */
    fun size(): Int {
        return trackList.size
    }

    /**
     * reloads song list
     */
    fun reload() {
        readSongsFromStorage()
    }

    companion object {

        val JSON_ARRAY_SONGS = "songs"
        val JSON_STRING_INTERPRET = "interpret"
        val JSON_STRING_TITLE = "title"
        val JSON_STRING_THUMB = "thumb"
        val JSON_LONG_DATE = "date"
    }
}